### PR TITLE
Refine joint decision sequential action mode

### PR DESCRIPTION
**Pull request by chaosddp**
# Description

Changes:
1. add a new mode: DecisionModel.JointWithSequentialAction
2. assign action to pending decision events from beginning to end, ignore events that have no action assigned under joint mode.

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [x] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/simulator/core.py [hunk 1]
```diff
@@ -286,31 +278,30 @@ def _simulate(self):
...
+                else:
+                    # For joint mode, we will assign actions from beginning to end.
+                    # Then mark others pending events to finished if not sequential action mode.
+                    for i, pending_event in enumerate(pending_events):
+                        if i >= len(actions):
+                            if self._decision_mode == DecisionMode.Joint:
+                                # Ignore following pending events that have no action matched.
+                                pending_event.state = EventState.FINISHED
+                        else:
+                            # Set the state as executing, so event buffer will not pop them again.
```
Comment (Jinyu-W) [inline maro/simulator/core.py:298]: pop them **again**?

File: maro/simulator/core.py [hunk 2]
```diff
@@ -286,31 +278,30 @@ def _simulate(self):
...
+                    # We just append the action into sub event of first pending cascade event.
+                    pending_events[0].state = EventState.EXECUTING
+                    pending_events[0].add_immediate_event(action_event, is_head=True)
+                else:
+                    # For joint mode, we will assign actions from beginning to end.
+                    # Then mark others pending events to finished if not sequential action mode.
+                    for i, pending_event in enumerate(pending_events):
+                        if i >= len(actions):
+                            if self._decision_mode == DecisionMode.Joint:
+                                # Ignore following pending events that have no action matched.
```
Comment (Jinyu-W) [inline maro/simulator/core.py:295]: So for the Joint Mode, if we have 5 decision events and receive only 3 actions, we map these 3 actions to the first 3 decision events? and directly mark treat the late 2 decision events as finished without any operation?

File: maro/simulator/core.py [hunk 3]
```diff
@@ -286,31 +278,30 @@ def _simulate(self):
...
-                    if is_support_seq_action:
-                        for i in range(1, pending_event_length):
-                            if pending_events[i].id == actions[0].src_event_id:
-                                pending_events[i].state = EventState.FINISHED
-                    else:
-                        for i in range(1, pending_event_length):
-                            pending_events[i].state = EventState.FINISHED
+                if self._decision_mode == DecisionMode.Sequential:
+                    # Generate a new atom event first.
+                    action_event = self._event_buffer.gen_action_event(self._tick, actions)
```
Comment (Jinyu-W) [inline maro/simulator/core.py:283]: for Sequential mode, the **actions** is a List?

**Review by Jinyu-W — COMMENTED** [review]
What situation is this mode for?

File: maro/simulator/core.py (hunk 1 above)
Comment (chaosddp) [reply maro/simulator/core.py:298]: same as joint mode, that pop all pending decision event at one time, not sequentially, it expect an action list that size is same as decision event. With this change, joint mode can accept actions one by one, and assign the to decision events in the same order, until all decision events get an action.

File: maro/simulator/core.py (hunk 3 above)
Comment (chaosddp) [reply maro/simulator/core.py:283]: just wrap it as a list, make it same with joint mode

File: maro/simulator/core.py (hunk 2 above)
Comment (chaosddp) [reply maro/simulator/core.py:295]: yes, joint mode expect an action list.

Comment (chaosddp) [reply maro/simulator/core.py:295]: And with new mode, we can pop the last 2 decision event again, as we will not mark them as finished.

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**